wnfs/src/private/hamt/node.rs

```diff
@@ -376,76 +377,81 @@ where
         }
     }
 
-    #[async_recursion(?Send)]
-    async fn remove_value<'a, 'b, B: BlockStore>(
-        self: &'a Rc<Self>,
-        hashnibbles: &'b mut HashNibbles,
-        store: &B,
-    ) -> Result<(Rc<Self>, Option<Pair<K, V>>)>
+    fn remove_value<'a, B: BlockStore>(
+        self: Rc<Self>,
+        hashnibbles: &'a mut HashNibbles,
+        store: &'a B,
+    ) -> LocalBoxFuture<'a, Result<RemovedResult<K, V, H>>>
     where
-        K: DeserializeOwned + Clone + AsRef<[u8]>,
-        V: DeserializeOwned + Clone,
+        K: DeserializeOwned + Clone + AsRef<[u8]> + 'a,
+        V: DeserializeOwned + Clone + 'a,
+        H: 'a,
     {
-        let bit_index = hashnibbles.try_next()?;
+        Box::pin(async move {
+            let bit_index = hashnibbles.try_next()?;
 
-        // If the bit is not set yet, return None.
-        if !self.bitmask[bit_index] {
-            return Ok((Rc::clone(self), None));
-        }
-
-        let value_index = self.get_value_index(bit_index);
-        Ok(match &self.pointers[value_index] {
-            Pointer::Values(values) => {
-                let mut node = (**self).clone();
-                let value = if values.len() == 1 {
-                    // If the key doesn't match, return without removing.
-                    if &H::hash(&values[0].key) != hashnibbles.digest {
-                        return Ok((Rc::clone(self), None));
-                    }
-                    // If there is only one value, we can remove the entire pointer.
-                    node.bitmask.set(bit_index, false);
-                    match node.pointers.remove(value_index) {
-                        Pointer::Values(mut values) => Some(values.pop().unwrap()),
-                        _ => unreachable!(),
-                    }
-                } else {
-                    // Otherwise, remove just the value.
-                    let mut values = (*values).clone();
-                    values
-                        .iter()
-                        .position(|p| &H::hash(&p.key) == hashnibbles.digest)
-                        .map(|i| {
-                            let value = values.remove(i);
-                            node.pointers[value_index] = Pointer::Values(values);
-                            value
-                        })
-                };
-
-                (Rc::new(node), value)
+            // If the bit is not set yet, return None.
+            if !self.bitmask[bit_index] {
```

Comment:
is this typically the bitarray under the hood? can you not use get, which exposes an option and use some option fns magic to return the result w/ a `?`. Always prefer the `?` over the early return.

Reply:
Actually, nevermind. There's lots of array bittwiddling happening in the HAMT implementation.
Any invalid access should be a show-stopping bug in the implementation.
I coded up something, but really didn't like it. Instead, I've added some tests to the deserialization process that makes sure you never decode something from the outside that has an invalid format. Thus, this should error out at the right point in time now.
I'm fairly confident the property based tests make sure we never actually break the invariants from within this implementation (i.e. we can skip bounds checks after we've validated the invariants from deserialized nodes).